strawberryfields/backends/tfbackend/ops.py

```diff
@@ -302,19 +304,22 @@ def grad(dy):  # pragma: no cover
     return gate, grad
 
 
-def displacement_matrix(r, phi, cutoff, batched=False):
+def displacement_matrix(r, phi, cutoff, batched=False, dtype=tf.complex64):
     """creates a single mode displacement matrix accounting for batching"""
-    r = tf.cast(r, def_type)
-    phi = tf.cast(phi, def_type)
+    r = tf.cast(r, dtype)
+    phi = tf.cast(phi, dtype)
     if batched:
         return tf.stack(
-            [single_displacement_matrix(r_, phi_, cutoff) for r_, phi_ in tf.transpose([r, phi])]
+            [
+                single_displacement_matrix(r_, phi_, cutoff, dtype=dtype.as_numpy_dtype)
+                for r_, phi_ in tf.transpose([r, phi])
+            ]
         )
-    return single_displacement_matrix(r, phi, cutoff)
+    return single_displacement_matrix(r, phi, cutoff, dtype=dtype.as_numpy_dtype)
```

Comment:
Is this because the `single_displacement_matrix` uses NumPy, and not tensorflow?

Reply:
They use Tensorflow to calculate the gradients, but TheWalrus to calculate matrix representations. So my interpretation is that TheWalrus uses NumPy.

Reply:
For some operations we use TW also for the gradients. And the implementation in TW itself uses numba.

Reply:
Yeah I think that's correct as far as TW. But numba uses numpy dtypes? That's why they get converted, as far as I know.

Reply:
yes, [we use numpy dtypes](https://github.com/XanaduAI/thewalrus/blob/f779cb91d57bd700aeb903fcb1c3132082a3f156/thewalrus/fock_gradients.py#L41). Then numba compiles the code to C upon first call.